src/execution_plan/execution_plan.c

```diff
@@ -527,39 +534,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
 	rm_free(plan);
 }
 
-// Free an op tree and its associated ExecutionPlan segments.
-static ExecutionPlan *_ExecutionPlan_FreeOpTree(OpBase *op) {
-	if(op == NULL) return NULL;
-	ExecutionPlan *child_plan = NULL;
-	ExecutionPlan *prev_child_plan = NULL;
-	// Store a reference to the current plan.
-	ExecutionPlan *current_plan = (ExecutionPlan *)op->plan;
+// free an op tree
+static void _ExecutionPlan_FreeOpTree(OpBase *op) {
+	if(op == NULL) {
+		return;
+	}
+
 	for(uint i = 0; i < op->childCount; i ++) {
-		child_plan = _ExecutionPlan_FreeOpTree(op->children[i]);
-		// In most cases all children will share the same plan, but if they don't
-		// (for an operation like UNION) then free the now-obsolete previous child plan.
-		if(prev_child_plan != child_plan && prev_child_plan != current_plan) {
-			_ExecutionPlan_FreeInternals(prev_child_plan);
-			prev_child_plan = child_plan;
-		}
+		_ExecutionPlan_FreeOpTree(op->children[i]);
 	}
 
 	// Free this op.
 	OpBase_Free(op);
+}
+
+// collect all ExecutionPlans from the given operation tree
+static void _ExecutionPlan_AggregatePlansFromOps
+(
+	OpBase *opBase,     // operation to aggregate plans from
+	rax *plans          // plans map to insert to
+) {
+	ASSERT(plans != NULL);
+
+	if(opBase == NULL) {
+		return;
+	}
 
-	// Free each ExecutionPlan segment once all ops associated with it have been freed.
-	if(current_plan != child_plan) _ExecutionPlan_FreeInternals(child_plan);
+	// add the plan if it doesn't exist already
+	static_assert(sizeof(opBase->plan) == sizeof(uint64_t), "pointer size is not 64 bits");
+	const uint64_t plan_ptr_be = _htobe64((uint64_t)opBase->plan);
+	raxTryInsert(plans, (unsigned char *)&plan_ptr_be, sizeof(uint64_t), (void *)opBase->plan, NULL);
```

Comment:
1. consider switching from `rax` to `HashTable`
2. consider using the following approach:
```c
OpFree(op) {
  op->plan->ops_count--;
  if(op->plan->ops_count == 0) {
    _ExecutionPlan_FreeInternals(op->plan);
  }
}
```